File: online-store-api/online-store-api/src/main/java/com/code81/onlinestore/service/CartService.java
package com.code81.onlinestore.service;

import com.code81.onlinestore.dto.cart.CartItemRequest;
import com.code81.onlinestore.dto.cart.CartItemUpdateRequest;
import com.code81.onlinestore.dto.cart.CartResponse;

public interface CartService {

    CartResponse getCart(Long customerId);

    CartResponse addItem(Long customerId, CartItemRequest request);

    CartResponse updateItem(Long customerId, Long productId, CartItemUpdateRequest request);

    CartResponse removeItem(Long customerId, Long productId);

    void clearCart(Long customerId);
}

File: code81_online_store_angular/src/app/core/services/cart.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { CartItemRequest, CartItemUpdateRequest, CartResponse } from '../models/cart.model';

@Injectable({ providedIn: 'root' })
export class CartService {
  private http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/cart`;

  getCart(): Observable<CartResponse> {
    return this.http.get<CartResponse>(this.baseUrl);
  }

  addItem(request: CartItemRequest): Observable<CartResponse> {
    return this.http.post<CartResponse>(`${this.baseUrl}/items`, request);
  }

  updateItem(productId: number, request: CartItemUpdateRequest): Observable<CartResponse> {
    return this.http.put<CartResponse>(`${this.baseUrl}/items/${productId}`, request);
  }

  removeItem(productId: number): Observable<CartResponse> {
    return this.http.delete<CartResponse>(`${this.baseUrl}/items/${productId}`);
  }

  clear(): Observable<void> {
    return this.http.delete<void>(this.baseUrl);
  }
}
